diff -ruN imap-2001a.orig/src/c-client/auth_md5.c imap-2001a/src/c-client/auth_md5.c
--- imap-2001a.orig/src/c-client/auth_md5.c	Wed May 30 22:42:31 2001
+++ imap-2001a/src/c-client/auth_md5.c	Thu Dec 13 13:59:09 2001
@@ -18,6 +18,7 @@
  * CPYRIGHT, included with this Distribution.
  */
 
+#include <time.h>
 /* MD5 context */
 
 #define MD5BLKLEN 64		/* MD5 block length */
diff -ruN imap-2001a.orig/src/osdep/unix/os_lnx.h imap-2001a/src/osdep/unix/os_lnx.h
--- imap-2001a.orig/src/osdep/unix/os_lnx.h	Tue Feb 20 17:45:54 2001
+++ imap-2001a/src/osdep/unix/os_lnx.h	Thu Dec 13 13:59:09 2001
@@ -18,6 +18,8 @@
  * CPYRIGHT, included with this Distribution.
  */
 
+#define _XOPEN_SOURCE 1
+#define _BSD_SOURCE 1
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
--- imap-2001a.orig/src/mlock/mlock.c	Thu Dec 27 10:56:54 2001
+++ imap-2001a/src/mlock/mlock.c	Thu Dec 27 10:55:52 2001
@@ -43,6 +43,7 @@
 #include <sys/file.h>
 #include <sys/stat.h>
 #include <sys/param.h>
+#include <stdlib.h>  /* malloc(3) */

 #define LOCKTIMEOUT 5          /* lock timeout in minutes */
 #define LOCKPROTECTION 0775

