diff -ruN imap-2003.DEV.SNAP-0303051641.orig/src/imapd/imapd.8c imap-2003.DEV.SNAP-0303051641/src/imapd/imapd.8c
--- imap-2003.DEV.SNAP-0303051641.orig/src/imapd/imapd.8c	1998-10-13 04:31:21.000000000 +0200
+++ imap-2003.DEV.SNAP-0303051641/src/imapd/imapd.8c	2003-03-08 20:22:30.000000000 +0100
@@ -3,7 +3,7 @@
 .SH NAME
 IMAPd \- Internet Message Access Protocol server
 .SH SYNOPSIS
-.B /usr/etc/imapd
+.B /usr/sbin/imapd
 .SH DESCRIPTION
 .I imapd
 is a server which supports the
@@ -26,7 +26,7 @@
 by many Unix-based clients.  To do this, the
 .I imapd
 binary must have a link to
-.I /etc/rimapd
+.I /usr/sbin/rimapd
 since this is where this software expects it to be located.
 .SH "SEE ALSO"
 rsh(1C) ipopd(8C)
diff -ruN imap-2003.DEV.SNAP-0303051641.orig/src/ipopd/ipopd.8c imap-2003.DEV.SNAP-0303051641/src/ipopd/ipopd.8c
--- imap-2003.DEV.SNAP-0303051641.orig/src/ipopd/ipopd.8c	1992-01-17 02:52:10.000000000 +0100
+++ imap-2003.DEV.SNAP-0303051641/src/ipopd/ipopd.8c	2003-03-08 20:22:30.000000000 +0100
@@ -3,9 +3,9 @@
 .SH NAME
 IPOPd \- Post Office Protocol server
 .SH SYNOPSIS
-.B /usr/etc/ipop2d
+.B /usr/sbin/ipop2d
 .PP
-.B /usr/etc/ipop3d
+.B /usr/sbin/ipop3d
 .SH DESCRIPTION
 .I IPOP2d
 and
@@ -28,6 +28,10 @@
 e.g.,
 .B SERVER.WASHINGTON.EDU:SMITH.
 .PP
+These daemons have been compiled with MD5 and APOP support.  See   
+.I /usr/share/doc/libc-client2001/md5.txt
+for additional information.
+.PP
 .I POPd
 is invoked by the internet server (see
 .IR inetd (8)),
diff -ruN imap-2003.DEV.SNAP-0303051641.orig/src/osdep/unix/Makefile imap-2003.DEV.SNAP-0303051641/src/osdep/unix/Makefile
--- imap-2003.DEV.SNAP-0303051641.orig/src/osdep/unix/Makefile	2003-03-08 20:21:31.000000000 +0100
+++ imap-2003.DEV.SNAP-0303051641/src/osdep/unix/Makefile	2003-03-08 20:28:24.000000000 +0100
@@ -74,7 +74,7 @@
 MAILSPOOL=$(SPOOLDIR)/mail
 NEWSSPOOL=$(SPOOLDIR)/news
 RSHPATH=/usr/ucb/rsh
-LOCKPGM=/etc/mlock
+LOCKPGM=/usr/bin/mlock
 
 
 # Default formats for creating new mailboxes and for empty mailboxes in the
diff -ruN imap-2003.DEV.SNAP-0303051641.orig/src/osdep/unix/tcp_unix.c imap-2003.DEV.SNAP-0303051641/src/osdep/unix/tcp_unix.c
--- imap-2003.DEV.SNAP-0303051641.orig/src/osdep/unix/tcp_unix.c	2002-10-31 21:04:41.000000000 +0100
+++ imap-2003.DEV.SNAP-0303051641/src/osdep/unix/tcp_unix.c	2003-03-08 20:22:31.000000000 +0100
@@ -323,12 +323,12 @@
 				/* return immediately if ssh disabled */
     if (!(sshpath && (ti = sshtimeout))) return NIL;
 				/* ssh command prototype defined yet? */
-    if (!sshcommand) sshcommand = cpystr ("%s %s -l %s exec /etc/r%sd");
+    if (!sshcommand) sshcommand = cpystr ("%s %s -l %s exec /usr/sbin/r%sd");
   }
   else if (ti = rshtimeout) {	/* set rsh timeout */
 				/* rsh path/command prototypes defined yet? */
     if (!rshpath) rshpath = cpystr (RSHPATH);
-    if (!rshcommand) rshcommand = cpystr ("%s %s -l %s exec /etc/r%sd");
+    if (!rshcommand) rshcommand = cpystr ("%s %s -l %s exec /usr/sbin/r%sd");
   }
   else return NIL;		/* rsh disabled */
 				/* look like domain literal? */
