--- imap-2004/src/c-client/auth_md5.c~	2005-01-17 16:38:46.758527958 -0700
+++ imap-2004/src/c-client/auth_md5.c	2005-01-17 16:38:46.758527958 -0700
@@ -153,7 +153,7 @@
 				/* get password */
       if (p = auth_md5_pwd ((authuser && *authuser) ? authuser : user)) {
 	pl = strlen (p);
-	u = (md5try && strcmp (hash,hmac_md5 (chal,cl,p,pl))) ? NIL : user;
+	u = (md5try && !strcmp (hash,hmac_md5 (chal,cl,p,pl))) ? user : NIL;
 	memset (p,0,pl);	/* erase sensitive information */
 	fs_give ((void **) &p);	/* flush erased password */
 				/* now log in for real */
