--- /usr/share/cdbs/1/rules/debhelper.mk	2005-03-10 13:35:06.000000000 +0100
+++ debhelper.mk	2005-04-03 09:30:18.000000000 +0200
@@ -183,7 +183,7 @@
 $(patsubst %,binary-makedeb/%,$(DEB_INDEP_REGULAR_PACKAGES)) :: binary-makedeb/% : common-binary-predeb-indep binary-makedeb-IMPL/%
 $(patsubst %,binary-makedeb-IMPL/%,$(DEB_PACKAGES)) :: binary-makedeb-IMPL/% : 
 	dh_shlibdeps -p$(cdbs_curpkg) $(DEB_DH_SHLIBDEPS_ARGS)
-	dh_gencontrol -p$(cdbs_curpkg) $(DEB_DH_GENCONTROL_ARGS)
+	dh_gencontrol -p$(cdbs_curpkg) $(DEB_DH_GENCONTROL_ARGS) $(if $(DEB_DH_GENCONTROL_DASHDASHARGS),-- $(DEB_DH_GENCONTROL_DASHDASHARGS))
 	dh_md5sums -p$(cdbs_curpkg) $(DEB_DH_MD5SUMS_ARGS)
 	$(DEB_DH_BUILDDEB_ENV) dh_builddeb -p$(cdbs_curpkg) $(DEB_DH_BUILDDEB_ARGS)
 
